### ISSUE TITLE
split models and commands APIs 

### ISSUE THREAD

**Issue by KostyaSha** [docker-java/docker-java#619]
Didn't found this as issue. Something that i suggested in comments. For discussion.

When this library initially existed it had request models separated from CMDs. That allowed to have CMD creation logic separately from models. It should also allow do serder on request models. Example could be http://download.eclipse.org/jgit/site/4.4.0.201606070830-r/apidocs/index.html with `org.eclipse.jgit.api  High-level API commands (the porcelain of JGit).` and other low level APIs. If somebody wants java docker client analog API, then ITs CMD, if somebody wants fill request object manually and send - "then low level".

Starting from 1.12 docker changes `POST /containers/(id or name)/start no longer accepts a HostConfig.` that mean that 
~~\- either we should stop supporting <1.10 (`"starting container with HostConfig was deprecated since v1.10 and removed in v1.12"}`).~~ 
- Either we should bundle more conditional logic. Keeping this logic in separate CMD without model imho sounds better.

PS. if API the same for 1.10-1.12 then i think we can ignore 1.9 version.

**Comment by KostyaSha** [comment]
@marcuslinke ?

**Comment by marcuslinke** [comment]
@KostyaSha Well, lets clarify whats means low-lewel and high-level in our case. In my oppinion  docker-java IS and MUST be a low-level API and not more. Any high-level API would combine the existing low-level CMDs to something new, maybe comparable to docker CLI. Maybe I have misunderstood the terms here... 

The second question is: What do you mean by splitting commands from request models? You can think of a CMD object as a (root) request model. It additionally combines the request model with the possible (http) parameters and it wouldn't make sense to decouple both. Sending JSON requests without these parameters is useless.

I understand your point with testablity of the request models and suggest to factor out the request models as base classes of the commands. So what we would get here for example is a `CreateContainerRequest` class that represents the request model. This one would be inherited by `CreateContainerCmd` which adds methods to set/get these additional (http) parameters and provides the `exec` method.

**Comment by KostyaSha** [comment]
> I understand your point with testablity of the request models and suggest to factor out the request models as base classes of the commands. So what we would get here for example is a CreateContainerRequest class that represents the request model. This one would be inherited by CreateContainerCmd which adds methods to set/get these additional (http) parameters and provides the exec method.

But we are in java and CMD is already inherits exec, using interfaces would be crazy i think.

> The second question is: What do you mean by splitting commands from request models? You can think of a CMD object as a (root) request model. It additionally combines the request model with the possible (http) parameters and it wouldn't make sense to decouple both. Sending JSON requests without these parameters is useless.

I mean doing model configuration not with CMD methods https://github.com/docker-java/docker-java/tree/docker-java-0.8.1#create-new-docker-container-wait-for-its-start-and-stop-it . Yes, you still need add some variables/headers in request, but they could be stored in CMD, or in model, or detected from model during exec.

> @KostyaSha Well, lets clarify whats means low-lewel and high-level in our case. In my oppinion docker-java IS and MUST be a low-level API and not more. Any high-level API would combine the existing low-level CMDs to something new, maybe comparable to docker CLI. Maybe I have misunderstood the terms here...

**Comment by marcuslinke** [comment]
> But we are in java and CMD is already inherits exec, using interfaces would be crazy i think.

The way I suggested would eleminate the CMD interfaces in package `com.github.dockerjava.api.command`. We don't really need interfaces for commands. 

> I mean doing model configuration not with CMD methods https://github.com/docker-java/docker-java/tree/docker-java-0.8.1#create-new-docker-container-wait-for-its-start-and-stop-it https://github.com/spotify/docker-client#usage-example . Yes, you still need add some variables/headers in request, but they could be stored in CMD, or in model, or detected from model during exec.

I don't see any advantage compared to what I've suggested here. It only complicates the usage from the user perspective. Passing a `CreateContainerRequest` object to a `CreateContainerCmd` is mandatory so why we should burden this to the API? So in fact this is a question of inheritance vs. composition where I would prefer inheritance because of the former reason here.

@KostyaSha Sadly you haven't clarified the meaning of the terms high-level and low-level from your perspective until now.

**Comment by KostyaSha** [comment]
> @KostyaSha Sadly you haven't clarified the meaning of the terms high-level and low-level from your perspective until now.

Highlevel it to have the same parameters with additional logic like docker tool has itself, it doesn't match to API directly. I.e. having run command, discussion that we had about networks. From low level perspective if HostConfig forbidden now, then we should just add new model objects and it will be user's issue to use right low level api. Now imho CMD bundled with request model is like low-leve+highlevel.

**Comment by marcuslinke** [comment]
As I stated before docker-java is low-level only. The fact we have introduced some high-level logic to commands is a workaround for a bad remote API design. We could implement high level API/commands on top of the existing low-level commands but this has nothing to do with this discussion here in my oppinion.

@KostyaSha So as I understand you want address different remote API versions with different request models? But what is with differences/changes in (html) parameters then? Letting the user choose which request model to use? This sounds like a bad idea. Supporting of multiple remote API versions is completely offtopic here is guess.
